Std/Data/ByteArray.lean

```diff
@@ -109,3 +109,21 @@ theorem get_extract_aux {a : ByteArray} {start stop} (h : i < (a.extract start s
 @[simp] theorem get_extract {a : ByteArray} {start stop} (h : i < (a.extract start stop).size) :
     (a.extract start stop)[i] = a[start+i]'(get_extract_aux h) := by
   simp [getElem_eq_data_getElem]
+
+/-! ### ofFn -/
+
+/--- `ofFn f` with `f : Fin n → UInt8` returns the byte array whose `i`th element is `f i`. --/
+def ofFn (f : Fin n → UInt8) : ByteArray where
+  data := .ofFn f
```

Comment:
This is an inefficient implementation, although I think we want this to be a defeq, so it should be overridden using `@[csimp]`.

Reply:
I'm glad you chimed in since I was wanted to ask you about this. I have a more efficient `csimp` version in the works but I paused it since I wasn't sure it was actually worthwhile. 

Both `ByteArray.mk` and `Array.ofFn` have efficient implementations. What ends up happening is that `Array.ofFn` builds an array boxing all the `UInt8`s, then `ByteArray.mk` (implemented in C as "lean_byte_array_mk") reads that array and unboxes all the `UInt8`'s. There's evident redundancy here but it's not clear to me that the one-pass version implemented in Lean would be that much faster in reality. (Of course, the gains in memory allocation are very clear.)

Since my primary usage is to use `ofFn` to make basic specs for some currently `opaque` defs, I don't currently have a need for an efficient version.

Anyway, I'm happy to continue working on my efficient `csimp` version, whether it ends up in this PR or a later one. I just wanted to consult the wizards before going further on this road!

Reply:
Yes, that's exactly the inefficiency I'm talking about. People expect functions like this to perform as well as can be expected under the circumstances, and this doesn't meet that bar. I would prefer to not merge this until this issue is addressed.

Reply:
Sounds good. Thanks!

Reply:
All done!